src/uct/ib/mlx5/ib_mlx5_md.c

```diff
@@ -0,0 +1,889 @@
+        }
+    } else if ((errno != EPERM) &&
+               (errno != EPROTONOSUPPORT) &&
+               (errno != EOPNOTSUPP)) {
+        ucs_error("MLX5_CMD_OP_QUERY_HCA_CAP failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err_free;
+    } else {
+        status = UCS_ERR_UNSUPPORTED;
+        goto err_free;
+    }
+
+    if (atomic) {
+        int ops = UCT_IB_MLX5_ATOMIC_OPS_CMP_SWAP |
+                  UCT_IB_MLX5_ATOMIC_OPS_FETCH_ADD;
+        uint8_t arg_size;
+        int cap_ops, mode8b;
+
+        UCT_IB_MLX5DV_SET(query_hca_cap_in, in, op_mod, UCT_IB_MLX5_HCA_CAP_OPMOD_GET_MAX |
+                                                       (UCT_IB_MLX5_CAP_ATOMIC << 1));
+        ret = mlx5dv_devx_general_cmd(ctx, in, sizeof(in), out, sizeof(out));
+        if (ret != 0) {
+            ucs_error("MLX5_CMD_OP_QUERY_HCA_CAP failed: %m");
+            return UCS_ERR_IO_ERROR;
+        }
+
+        arg_size = UCT_IB_MLX5DV_GET(atomic_caps, cap, atomic_size_qp);
+        cap_ops  = UCT_IB_MLX5DV_GET(atomic_caps, cap, atomic_operations);
+        mode8b   = UCT_IB_MLX5DV_GET(atomic_caps, cap, atomic_req_8B_endianness_mode);
+
+        if ((cap_ops & ops) == ops) {
+            dev->atomic_arg_sizes = sizeof(uint64_t);
+            if (!mode8b) {
+                dev->atomic_arg_sizes_be = sizeof(uint64_t);
+            }
+        }
+
+        ops |= UCT_IB_MLX5_ATOMIC_OPS_MASKED_CMP_SWAP |
+               UCT_IB_MLX5_ATOMIC_OPS_MASKED_FETCH_ADD;
+
+        arg_size &= UCT_IB_MLX5DV_GET(query_hca_cap_out, out,
+                                      capability.atomic_caps.atomic_size_dc);
+
+        if ((cap_ops & ops) == ops) {
+            dev->ext_atomic_arg_sizes = arg_size;
+            if (mode8b) {
+                arg_size &= ~(sizeof(uint64_t));
+            }
+            dev->ext_atomic_arg_sizes_be = arg_size;
+        }
+
+        dev->pci_fadd_arg_sizes  = UCT_IB_MLX5DV_GET(atomic_caps, cap, fetch_add_pci_atomic) << 2;
+        dev->pci_cswap_arg_sizes = UCT_IB_MLX5DV_GET(atomic_caps, cap, compare_swap_pci_atomic) << 2;
+    }
+
+    md->super.ops = &uct_ib_mlx5dv_md_ops;
+    status = uct_ib_md_open_common(&md->super, ibv_device, md_config);
+    if (status != UCS_OK) {
+        goto err_free;
+    }
+
+    status = ucs_mpool_init(&md->dbrec_pool, 0,
+                            sizeof(uct_ib_mlx5_dbrec_t), 0,
+                            UCS_SYS_CACHE_LINE_SIZE,
+                            ucs_get_page_size() / UCS_SYS_CACHE_LINE_SIZE - 1,
+                            UINT_MAX, &uct_ib_mlx5_dbrec_ops, "devx dbrec");
+    if (status != UCS_OK) {
+        goto err_free;
+    }
+
+    dev->flags |= UCT_IB_DEVICE_FLAG_MLX5_PRM;
+    md->flags |= UCT_IB_MLX5_MD_FLAG_DEVX;
+    *p_md = &md->super;
+    return status;
+
+err_free:
+    ucs_free(md);
+err_free_context:
+    ibv_close_device(ctx);
+err:
+    return status;
+}
+
+#else
+
+static ucs_status_t uct_ib_mlx5_check_dc(uct_ib_device_t *dev)
```

Comment:
uct_ib_mlx5dv_check_dc

Reply:
unresolved

Reply:
but then you saw it's not really mlx5dv...

Reply:
right..

Reply:
maybe add a comment for now - "check dc support for mlx5dv and exp verbs"